Consolidate all the Error Message types (#1159)
moq-wg/moq-transport: draft-ietf-moq-transport.md

|-------|-----------------------------------------------------|
| 0x1A  | REQUESTS_BLOCKED ({{message-requests-blocked}})     |
|-------|-----------------------------------------------------|

> | 0x2   | REQUEST_ERROR  ({{message-request-ok}})             |

Reviewer: 0x2 is SUBSCRIBE_UPDATE

Please pick a new unused codepoint.

* Maximum Request ID: The Maximum Request ID for the session on which the
  endpoint is blocked. More on Request ID in {{request-id}}.

## REQUEST_ERROR {#message-request-error}

The REQUEST_ERROR message is sent to a response to any request (SUBSCRIBE, FETCH,
PUBLISH, SUBSCRIBE_NAMESPACE, PUBLISH_NAMESPACE, TRACK_STATUS). The unique
request ID in the REQUEST_ERROR is used to associate it with the correct type of
request.

~~~
REQUEST_ERROR Message {
  Type (i) = 0x5,
  Length (16),
  Request ID (i),
  Error Code (i),
  Error Reason (Reason Phrase),
}
~~~
{: #moq-transport-request-error format title="MOQT REQUEST_ERROR Message"}

* Request ID: The Request ID to which this message is replying.

* Error Code: Identifies an integer error code for request failure.

* Error Reason: Provides a text description of the request error. See
 {{reason-phrase}}.

The application SHOULD use a relevant error code in REQUEST_ERROR,
as defined below. Most codepoints have identical meanings for various request
types, but some have request-specific meanings.

INTERNAL_ERROR (0x0):

> : An implementation specific or generic error occurred.

Reviewer: @fluffy I understand the sentiment, but I disagree that _this_ PR should be blocked on resolving every individual error code semantic. That is completely orthogonal to what is happening here.

UNAUTHORIZED (0x1):
: The subscriber is not authorized to subscribe to the given track.

TIMEOUT (0x2):
: The subscription could not be completed before an implementation specific
  timeout. For example, a relay could not establish an upstream subscription
  within the timeout.

NOT_SUPPORTED (0x3):
: The endpoint does not support the type of request.

DOES_NOT_EXIST or UNINTERESTED (0x4):
: In response to SUBSCRIBE, FETCH, TRACK_STATUS, or SUBSCRIBE_NAMESPACE, the
track or namespace is not available at the publisher. For PUBLISH or
PUBLISH_NAMESPACE, the subscriber is not interested in the track or namespace.

INVALID_RANGE or PREFIX_OVERLAP (0x5):
: In response to SUBSCRIBE or FETCH, specified range of Locations cannot be
accessed with the request type. In response to SUBSCRIBE_NAMESPACE, the
namespace prefix overlaps with another SUBSCRIBE_NAMESPACE in the same
session.

NO_OBJECTS (0x6):

> : In response to FETCH, no objects exist between the Start and End Locations.

Reviewer 2: This is not really an error and I thought we decided to just sent a fetch response with no data in it.

INVALID_JOINING_REQUEST_ID(0x7):
: In response to a Joining FETCH, the referenced Request ID is not an active
Subscription.

UNKNOWN_STATUS_IN_RANGE(0x8):
: In response to a FETCH, the requested range contains an object with unknown
status.

MALFORMED_TRACK (0x9):
: In response to a SUBSCRIBE, FETCH, or TRACK_STATUS, a relay publisher detected
the track was malformed (see {{malformed-tracks}}).

MALFORMED_AUTH_TOKEN (0x10):
: Invalid Auth Token serialization during registration (see

>   {{authorization-token}}).

Reviewer 2: Seems like we need both a Malformed and Invalid auth token error. Name does not match description.

* Maximum Request ID: The Maximum Request ID for the session on which the
  endpoint is blocked. More on Request ID in {{request-id}}.

## REQUEST_ERROR {#message-request-error}

The REQUEST_ERROR message is sent to a response to any request (SUBSCRIBE, FETCH,
PUBLISH, SUBSCRIBE_NAMESPACE, PUBLISH_NAMESPACE, TRACK_STATUS). The unique
request ID in the REQUEST_ERROR is used to associate it with the correct type of
request.

~~~
REQUEST_ERROR Message {
  Type (i) = 0x5,
  Length (16),
  Request ID (i),
  Error Code (i),
  Error Reason (Reason Phrase),
}
~~~
{: #moq-transport-request-error format title="MOQT REQUEST_ERROR Message"}

* Request ID: The Request ID to which this message is replying.

* Error Code: Identifies an integer error code for request failure.

* Error Reason: Provides a text description of the request error. See
 {{reason-phrase}}.

The application SHOULD use a relevant error code in REQUEST_ERROR,
as defined below. Most codepoints have identical meanings for various request
types, but some have request-specific meanings.

INTERNAL_ERROR (0x0):
: An implementation specific or generic error occurred.

UNAUTHORIZED (0x1):

> : The subscriber is not authorized to subscribe to the given track.

Reviewer 3: ```suggestion
: The subscriber is not authorized to perform the requested action on the given track.
```

TIMEOUT (0x2):
: The subscription could not be completed before an implementation specific
  timeout. For example, a relay could not establish an upstream subscription
  within the timeout.

NOT_SUPPORTED (0x3):

> : The endpoint does not support the type of request.

Reviewer: Isn't "processing the request" implied since it's in the error message?  Which other endpoint would it be?

MALFORMED_AUTH_TOKEN (0x4):
: Invalid Auth Token serialization during registration (see
  {{authorization-token}}).

EXPIRED_AUTH_TOKEN (0x5):
: Authorization token has expired ({{authorization-token}}).

Below are errors for use by the publisher. They can appear in response to

> SUBSCRIBE, FETCH, TRACK_STATUS, and SUBSCRIBE_NAMESPACE, unless otherwise noted.

Reviewer: For that, we need to resolve #1175 - I expect they would be treated the same way?

DOES_NOT_EXIST (0x10):
: The track or namespace is not available at the publisher.

INVALID_RANGE (0x11):

> : In response to SUBSCRIBE or FETCH, specified range of Locations cannot be
> accessed with the request type.

Reviewer 3: ```suggestion
accessed with the request type.
```
```suggestion
: In response to SUBSCRIBE or FETCH, specified range of Locations or Filters in the request, cannot be
satisfied.
```

MALFORMED_TRACK (0x12):

> : In response to a SUBSCRIBE, FETCH, or TRACK_STATUS, a relay publisher detected

Reviewer: I repurposed #1158 to discuss the cacheability of malformed track which impacts whether subscribe or track status can result in a malformed track.  I recommend keeping this constrained to FETCH until that resolves.
```suggestion
: In response to a FETCH, a relay publisher detected
```

the track was malformed (see {{malformed-tracks}}).

The following are errors for use by the subscriber. They can appear in response
to PUBLISH or PUBLISH_NAMESPACE, unless otherwise noted.

UNINTERESTED (0x20):

> : The subscriber is not interested in the track or namespace.

Reviewer 3: Under what conditions, a publisher chooses to use this error. This error seems to be not much helpful, unless I am missing something

Reviewer: This was already discussed in #704.

A publisher sends a SUBSCRIBE_ERROR control message in response to a
failed SUBSCRIBE.

~~~
SUBSCRIBE_ERROR Message {
  Type (i) = 0x5,
  Length (16),
  Request ID (i),
  Error Code (i),
  Error Reason (Reason Phrase),
}
~~~
{: #moq-transport-subscribe-error format title="MOQT SUBSCRIBE_ERROR Message"}

* Request ID: The Request ID of the SUBSCRIBE this message is replying to
  {{message-subscribe-req}}.

* Error Code: Identifies an integer error code for subscription failure.

* Error Reason: Provides the reason for subscription error. See {{reason-phrase}}.

The application SHOULD use a relevant error code in SUBSCRIBE_ERROR,
as defined below:

INTERNAL_ERROR (0x0):
: An implementation specific or generic error occurred.

UNAUTHORIZED (0x1):
: The subscriber is not authorized to subscribe to the given track.

TIMEOUT (0x2):
: The subscription could not be completed before an implementation specific
  timeout. For example, a relay could not establish an upstream subscription
  within the timeout.

NOT_SUPPORTED (0x3):
: The endpoint does not support the SUBSCRIBE method.

TRACK_DOES_NOT_EXIST (0x4):
: The requested track is not available at the publisher.

> 

Reviewer: Now called `DOES_NOT_EXIST`, see line 1908

* Parameters: The parameters are defined in {{version-specific-params}}.

## FETCH_ERROR {#message-fetch-error}

A publisher sends a FETCH_ERROR control message in response to a
failed FETCH.

~~~
FETCH_ERROR Message {
  Type (i) = 0x19,
  Length (16),
  Request ID (i),
  Error Code (i),
  Error Reason (Reason Phrase)
}
~~~
{: #moq-transport-fetch-error format title="MOQT FETCH_ERROR Message"}

* Request ID: The Request ID of the FETCH this message is replying to
  {{message-subscribe-req}}.

* Error Code: Identifies an integer error code for fetch failure.

* Error Reason: Provides the reason for fetch error. See {{reason-phrase}}.

The application SHOULD use a relevant error code in FETCH_ERROR,
as defined below:

INTERNAL_ERROR (0x0):
: An implementation specific or generic error occurred.

UNAUTHORIZED (0x1):
: The subscriber is not authorized to fetch from the given track.

TIMEOUT (0x2):
: The fetch could not be completed before an implementation specific timeout.
  For example, a relay could not FETCH missing objects within the timeout.

NOT_SUPPORTED (0x3):

> : The endpoint does not support the FETCH method.

Reviewer: See line 1895

TRACK_DOES_NOT_EXIST (0x4):

> : The requested track is not available at the publisher.

Reviewer: Now called `DOES_NOT_EXIST`; see line 1908.

## SUBSCRIBE_NAMESPACE_ERROR {#message-sub-ns-error}

A publisher sends a SUBSCRIBE_NAMESPACE_ERROR control message in response to
a failed SUBSCRIBE_NAMESPACE.

~~~
SUBSCRIBE_NAMESPACE_ERROR Message {
  Type (i) = 0x13,
  Length (16),
  Request ID (i),
  Error Code (i),
  Error Reason (Reason Phrase)
}
~~~
{: #moq-transport-sub-ann-error format
title="MOQT SUBSCRIBE_NAMESPACE_ERROR Message"}

* Request ID: The Request ID of the SUBSCRIBE_NAMESPACE this message is replying
  to {{message-subscribe-ns}}.

* Error Code: Identifies an integer error code for the namespace subscription
failure.

* Error Reason: Provides the reason for the namespace subscription error.
  See {{reason-phrase}}.

The application SHOULD use a relevant error code in SUBSCRIBE_NAMESPACE_ERROR,
as defined below:

INTERNAL_ERROR (0x0):
: An implementation specific or generic error occurred.

UNAUTHORIZED (0x1):
: The subscriber is not authorized to subscribe to the given namespace prefix.

TIMEOUT (0x2):
: The operation could not be completed before an implementation specific
  timeout.

NOT_SUPPORTED (0x3):
: The endpoint does not support the SUBSCRIBE_NAMESPACE method.

NAMESPACE_PREFIX_UNKNOWN (0x4):

> : The namespace prefix is not available for subscription.

Reviewer 3: we are missing these error codes .

Reviewer: I believe this PR is consolidating this error with `DOES_NOT_EXIST`, is that right @martinduke?

NAMESPACE_PREFIX_OVERLAP (0x5):
: The namespace prefix overlaps with another SUBSCRIBE_NAMESPACE in the same

>   session.

Reviewer 3: same as above .

Reviewer: Line 1927

| INVALID_AUTHORITY          | 0x19 | {{session-termination}} |
| MALFORMED_AUTHORITY        | 0x1A | {{session-termination}} |

### REQUEST_ERROR Codes {#iana-request-error}

> | Name                       | Code | Request Types  |Specification              |

Author: done, though I didn't bother to create separate tables

| INTERNAL_ERROR             | 0x0  | All            | {{message-request-error}} |
| UNAUTHORIZED               | 0x1  | All            | {{message-request-error}} |
| TIMEOUT                    | 0x2  | All            | {{message-request-error}} |
| NOT_SUPPORTED              | 0x3  | All            | {{message-request-error}} |
| DOES_NOT_EXIST             | 0x4  | SUBSCRIBE, FETCH, TRACK_STATUS, SUBSCRIBE_NAMESPACE | {{message-request-error}} |
| UINTERESTED                | 0x4  | PUBLISH, PUBLISH_NAMESPACE | {{message-request-error}} |
| INVALID_RANGE              | 0x5  | SUBSCRIBE, FETCH | {{message-request-error}} |
| PREFIX_OVERLAP             | 0x5  | SUBSCRIBE_NAMESPACE | {{message-request-error}} |
| NO_OBJECTS                 | 0x6  | FETCH          | {{message-request-error}} |
| INVALID_JOINING_REQUEST_ID | 0x7  | FETCH          | {{message-request-error}} |
| UNKNOWN_STATUS_IN_RANGE    | 0x8  | FETCH          | {{message-request-error}} |
| MALFORMED_TRACK            | 0x9  | SUBSCRIBE, FETCH, TRACK_STATUS | {{message-request-error}} |
| MALFORMED_AUTH_TOKEN       | 0x10 | All            | {{message-request-error}} |

> | EXPIRED_AUTH_TOKEN         | 0x12 | All            | {{message-request-error}} |

Reviewer: I think we could do something interesting with the ranges here, all the common errors in the 0-X range, generic subscriber, generic publisher, then message specific error ranges, etc.

Reviewer: > Error are only really usefull if the receiver can do something when they receive them.

Strong disagree.  In practice, errors also feed monitoring systems that have alarms and operators behind them.  Even if every error had the same client action "give up" or "retry", being able to group these by the actual problem has a lot of value.

> I don't think class of errors ever helped even in HTTP.

As someone who operates an HTTP CDN at scale, I will say that error classes are very useful, for the above reason, and you can look at any HTTP implementation and I virtually guarantee you will find code that divides status by 100 and looks at the result.

That said,my suggestion to band the error codes had less to do with the way that HTTP does it (by function), but instead just grouping them by verb (subscribe specific, publish specific, etc).
